pkg/utils/metrics/setup.go

```diff
@@ -10,11 +10,37 @@ var (
 		Name: "aws_api_calls_total",
 		Help: "Number of API calls to the AWS API",
 	}, []string{"service", "operation", "api_version"})
+	// MetricAWSAPICallsRec can be used in controllers for exposing a total amount of AWS API calls for every create/observe/update/delete operation
+	MetricAWSAPICallsRec = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "aws_api_calls_reconcile_managed_resource_total",
+		Help: "Amount of calls to the AWS API produced by controller per reconciliation for every managed resource and controller operation type",
+	}, []string{"api_version", "kind", "resource_name", "controller_operation_type"})
+	// MetricManagedResRec can be used in controllers for exposing a total amount of reconciliations for every managed resource
+	MetricManagedResRec = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "managed_resource_reconciles_total",
+		Help: "Total amount of reconciliation loops per managed resource",
+	}, []string{"api_version", "kind", "resource_name"})
```

Comment:
I'm not so sure about these additional metrics. Are they really necessary?

If this would be applied to every controller, it would create a huge number of additional metrics: `189 CRDs x 4 x #MRs`.

If it is not actually necessary I would propose to leave it out of this PR to get it merged and maybe open a second one for the metrics.

Reply:
@MisterMX thanks for looking at the pr!

>I'm not so sure about these additional metrics. Are they really necessary?
Well, it depends. I think this question is actually for a major part of 6600+ metrics which are already exposed, including `aws_api_calls_total`.
I believe it is nice to have, because based on these metrics you can precisely observe the amount of calls for each of your managed resources, it helps find 'right' poll interval and understand your aws cost.

>If this would be applied to every controller, it would create a huge number of additional metrics: 189 CRDs x 4 x #MRs.

Not exactly: )
Unlike metricAWSAPICalls(aws_api_calls_total) these two new counters should be implemented `manually` for every controller, because there is no way(or I didn't find it) to use middleware for automated exposing. No data about managed resources and reconciles on the client scope.
So these metrics work only for ProvisioneProduct for now, and only if it is used, I mean these metrics are exposed per managed resource.

 As I texted above, the metrics exposed per managed resource(ProvisionedProduct) but not every service CRD. But if so(it is not), it is okay for prom endpoints. Current payload of the `/metrics` endpoint is around ~650K which is not small but okay and 400+ metrics doesn't make a big change. But again, it is not 400+. 

>If it is not actually necessary I would propose to leave it out of this PR to get it merged and maybe open a second one for the metrics.

I also thought about it, but these two metrics work only if implemented on the controller side, that is why I decided to put everything in this PR. Cuz I didn't touch other/existeng resources, it's a lot of work and testing.



Ps: I found inaccuracy in the description of `aws_api_calls_reconcile_managed_resource_total` - it is total counter but not `per reconciliation`. I will fix it if my arguments have dispelled your concerns.

Reply:
Thanks for the clarification. If I understand it right if it would be implemented for every controller we would still most likely end up with at least 4 metrics per MR? Given that users could have thousands of MRs this would mean a huge number of additional metrics (on top of the already huge amount).

I understand why you want this for this particular controller. However, I would like to have the metrics part to be implemented and discussed separately from this PR. I am not opposed to have additional and more detailed metrics for controllers. But there are some more things that should be considered in this discussion, like the implementation for other controllers and the amount of metrics being created. Ideally, we should also give users the ability to control which metrics should be exposed so we are not creating to much spam.

Reply:
>least 4 metrics per MR?

Even more - 5
```
aws_api_calls_reconcile_managed_resource_total{api_version="servicecatalog.aws.crossplane.io/v1alpha1",controller_operation_type="create",kind="ProvisionedProduct",resource_name="kirills-test-2"} 1
aws_api_calls_reconcile_managed_resource_total{api_version="servicecatalog.aws.crossplane.io/v1alpha1",controller_operation_type="delete",kind="ProvisionedProduct",resource_name="kirills-test-2"} 0
aws_api_calls_reconcile_managed_resource_total{api_version="servicecatalog.aws.crossplane.io/v1alpha1",controller_operation_type="observe",kind="ProvisionedProduct",resource_name="kirills-test-2"} 27415
aws_api_calls_reconcile_managed_resource_total{api_version="servicecatalog.aws.crossplane.io/v1alpha1",controller_operation_type="update",kind="ProvisionedProduct",resource_name="kirills-test-2"} 0
managed_resource_reconciles_total{api_version="servicecatalog.aws.crossplane.io/v1alpha1",kind="ProvisionedProduct",resource_name="kirills-test-2"} 5483
```

>However, I would like to have the metrics part to be implemented and discussed separately from this PR
...
Ideally, we should also give users the ability to control which metrics should be exposed so we are not creating to much spam.

Okay, I will move metrics to dedicated branch and then create a pr, also I will think about feature flag or something. But first I'll cleanup this branch, it will take some time. Probably tomorow it will be ready.

Reply:
@MisterMX it is done - I removed metrics from this PR.

Reply:
@MisterMX could you run ci checks once again?
Idk what is wrong, [here](https://github.com/swisscom/provider-aws/actions/runs/6577103192) and localy everything is okay.

Reply:
@teeverr some refactoring has been made in #1917. The code in `pkg/clients` has been moved to `pkg/utils`. You need to rebase on the latest master.

Reply:
Oh, I see. Thanks!
I rebased it with the latest master